src/audio/codec_adapter/codec_adapter.c

```diff
@@ -399,84 +399,39 @@ int codec_adapter_copy(struct comp_dev *dev)
-	comp_dbg(dev, "codec_adapter_set_params(): start: num_of_elem %d, elem remain %d msg_index %u",
+	comp_dbg(dev, "codec_adapter_set_params(): num_of_elem %d, elem remain %d msg_index %u",
 		 cdata->num_elems, cdata->elems_remaining, cdata->msg_index);
 
-	/* Stage 1 verify input params & allocate memory for the config blob */
-	if (cdata->msg_index == 0) {
+	/* set the fragment position, data offset and config data size */
+	if (!cdata->msg_index) {
 		size = cdata->num_elems + cdata->elems_remaining;
-		/* Check that there is no work-in-progress on previous request */
-		if (md->runtime_params) {
-			comp_err(dev, "codec_adapter_set_params() error: busy with previous request");
-			return -EBUSY;
-		}
-
-		if (!size) {
-			comp_err(dev, "codec_adapter_set_params() error: no configuration size %d",
-				 size);
-			/* TODO: return -EINVAL. This is temporary until driver fixes its issue */
-			return 0;
-		}
-
-		if (size > MAX_BLOB_SIZE) {
-			comp_err(dev, "codec_adapter_set_params() error: blob size is too big cfg size %d, allowed %d",
-				 size, MAX_BLOB_SIZE);
-			return -EINVAL;
-		}
-
-		/* Allocate buffer for new params */
-		md->runtime_params = rballoc(0, SOF_MEM_CAPS_RAM, size);
-		if (!md->runtime_params) {
-			comp_err(dev, "codec_adapter_set_params(): space allocation for new params failed");
-			return -ENOMEM;
-		}
-
-		memset(md->runtime_params, 0, size);
-	} else if (!md->runtime_params) {
-		comp_err(dev, "codec_adapter_set_params() error: no memory available for runtime params in consecutive load");
-		return -EIO;
-	}
-
-	offset = size - (cdata->num_elems + cdata->elems_remaining);
-	dst = (char *)md->runtime_params + offset;
-	src = (char *)cdata->data->data;
-
-	ret = memcpy_s(dst, size - offset, src, cdata->num_elems);
-	assert(!ret);
-
-	/* return as more fragments of config data expected */
-	if (cdata->elems_remaining)
-		return 0;
-
-	/* config fully copied, now load it */
-	ret = module_load_config(dev, md->runtime_params, size);
-	if (ret)
-		goto end;
-
-	comp_dbg(dev, "codec_adapter_set_params(): config load successful");
-
-	/* And apply runtime config right away if codec is already prepared */
-	if (md->state >= MODULE_INITIALIZED) {
-		ret = module_apply_runtime_config(dev);
-		if (ret)
-			comp_err(dev, "codec_adapter_set_params() error %x: codec runtime config apply failed",
-				 ret);
+		data_offset_size = size;
+		if (!cdata->elems_remaining)
+			pos = MODULE_CFG_FRAGMENT_SINGLE;
+		else
+			pos = MODULE_CFG_FRAGMENT_FIRST;
+	} else {
+		data_offset_size = size - (cdata->num_elems + cdata->elems_remaining);
+		if (!cdata->elems_remaining)
+			pos = MODULE_CFG_FRAGMENT_LAST;
 		else
-			comp_dbg(dev, "codec_adapter_set_params() apply of runtime config done.");
+			pos = MODULE_CFG_FRAGMENT_MIDDLE;
 	}
 
-end:
-	if (md->runtime_params)
-		rfree(md->runtime_params);
-	md->runtime_params = NULL;
-	return ret;
+	if (md->ops->set_configuration)
+		return md->ops->set_configuration(mod, 0, pos, data_offset_size,
```

Comment:
The `config_id `field is required for ipc4 compatibility, we cannot just pass 0.

Reply:
but this is only with IPC3 that it will be 0. With IPC4, it will be the actual ID.

Reply:
Since the new Module API is driven by IPC4 introduction we should make sure that proposed changes are compliant with it. I would like to avoid doing major Module API changes multiple times in a short period of time.

Reply:
But this wouldn't require any API changes. In IPC3, we just need the config_id. When this api is used with IPC4, we will have the correct config_id to pass.